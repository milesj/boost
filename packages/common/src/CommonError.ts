import { createScopedError } from '@boost/internal';

const errors = {
	PARSE_INVALID_EXT: 'Unable to parse file "{0}". Unsupported file extension.',
	PATH_REQUIRE_ABSOLUTE: 'An absolute file path is required.',
	PATH_RESOLVE_LOOKUPS: 'Failed to resolve a path using the following lookups (in order):\n{0}\n',
	PROJECT_NO_PACKAGE: 'No `package.json` found within project root.',
};

export type CommonErrorCode = keyof typeof errors;

export default createScopedError<CommonErrorCode>('CMN', 'CommonError', errors);
