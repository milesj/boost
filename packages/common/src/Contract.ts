import optimal, { Blueprint, Predicates, predicates } from 'optimal';
import { Optionable } from './types';

export default abstract class Contract<T extends object = {}> implements Optionable<T> {
  readonly options: Readonly<Required<T>>;

  constructor(options?: T) {
    this.options = this.configure(options);
  }

  /**
   * Set an options object by merging the new partial and existing options
   * with the defined blueprint, while running all validation checks.
   * Freeze and return the options object.
   */
  configure(options?: Partial<T> | ((options: Required<T>) => Partial<T>)): Readonly<Required<T>> {
    const nextOptions = typeof options === 'function' ? options(this.options) : options;

    // We don't want the options property to be modified directly,
    // so it's read only, but we still want to modify it with this function.
    // @ts-expect-error
    this.options = Object.freeze(
      optimal(
        { ...this.options, ...nextOptions },
        this.blueprint(predicates, this.options === undefined) as Blueprint<T>,
        {
          name: this.constructor.name,
        },
      ),
    );

    return this.options;
  }

  /**
   * Define an optimal blueprint in which to validate and build the
   * options object passed to the constructor, or when manual setting.
   */
  abstract blueprint(predicates: Predicates, onConstruction?: boolean): Blueprint<object>;
}
