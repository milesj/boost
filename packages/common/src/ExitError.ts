export class ExitError extends Error {
	code: number;

	constructor(message: string, code: number) {
		super(message);

		this.code = code;
		this.name = 'ExitError';
	}
}
