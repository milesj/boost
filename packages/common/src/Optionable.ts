import optimal, { predicates, Blueprint, Predicates } from 'optimal';

export default abstract class Optionable<T extends object = {}> {
  options: Required<T>;

  constructor(options?: T) {
    this.options = this.doConfigure(options);
  }

  /**
   * Validate and build the options object using optimal.
   */
  configure(options: Partial<T>): this {
    this.options = this.doConfigure(options);

    return this;
  }

  /**
   * Protected helper to execute optimal.
   */
  private doConfigure(options?: Partial<T>): Required<T> {
    return optimal(
      {
        ...this.options,
        ...options,
      },
      this.blueprint(predicates),
      {
        name: this.constructor.name,
      },
    );
  }

  /**
   * Define an optimal blueprint in which to validate and build the
   * options object passed to the constructor.
   */
  abstract blueprint(predicates: Predicates): Blueprint<T>;
}
