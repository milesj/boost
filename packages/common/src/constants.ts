// https://github.com/npm/validate-npm-package-name
export const MODULE_NAME_PART = /[a-z0-9]{1}[-a-z0-9_.]*/u;

export const MODULE_NAME_PATTERN = new RegExp(
	`^(?:@(${MODULE_NAME_PART.source})/)?(${MODULE_NAME_PART.source})$`,
	'u',
);
