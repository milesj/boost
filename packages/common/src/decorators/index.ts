export { default as Bind } from './Bind';
export { default as Debounce } from './Debounce';
export { default as Deprecate } from './Deprecate';
