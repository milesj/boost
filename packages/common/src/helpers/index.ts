export * from './createBlueprint';
export * from './deepFreeze';
export * from './deepMerge';
export * from './formatMs';
export * from './instanceOf';
export * from './isEmpty';
export * from './isFilePath';
export * from './isModuleName';
export * from './isObject';
export * from './isPlainObject';
export * from './toArray';
