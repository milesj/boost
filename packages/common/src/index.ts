/**
 * @copyright   2020, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import optimal, { predicates, Blueprint, Predicates } from 'optimal';
import CommonError, { CommonErrorCode } from './CommonError';
import Contract from './Contract';
import ExitError from './ExitError';
import Path from './Path';
import PathResolver from './PathResolver';
import Project from './Project';
import * as json from './serializers/json';
import * as yaml from './serializers/yaml';

export * from './constants';
export * from './helpers';
export * from './types';

export {
  CommonError,
  CommonErrorCode,
  Contract,
  Blueprint,
  ExitError,
  Path,
  PathResolver,
  Predicates,
  Project,
  optimal,
  predicates,
  json,
  yaml,
};
