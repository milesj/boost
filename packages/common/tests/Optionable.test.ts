import Optionable from '../src/Optionable';
import { Predicates } from '../src';

describe('Optionable', () => {
  class OptionalProps extends Optionable<{ foo?: string; bar?: number }> {
    blueprint({ number, string }: Predicates) {
      return {
        foo: string('default'),
        bar: number(),
      };
    }
  }

  class RequiredProps extends Optionable<{ bar: number }> {
    // eslint-disable-next-line no-useless-constructor
    constructor(options: { bar: number }) {
      super(options);
    }

    blueprint({ number }: Predicates) {
      return {
        bar: number().required(),
      };
    }
  }

  describe('constructor()', () => {
    it('errors for invalid option type passed', () => {
      expect(
        () =>
          // @ts-ignore Allow invalid type
          new OptionalProps({
            foo: 123,
          }),
      ).toThrowErrorMatchingSnapshot();
    });

    it('errors for unknown option', () => {
      expect(
        () =>
          new OptionalProps({
            // @ts-ignore Allow unknown
            unknown: 'abc',
          }),
      ).toThrowErrorMatchingSnapshot();
    });

    it('sets options', () => {
      const opts = new OptionalProps({
        foo: 'custom',
      });

      expect(opts.options.foo).toBe('custom');
    });

    it('inherits option defaults', () => {
      const opts = new OptionalProps();

      expect(opts.options.foo).toBe('default');
    });

    it('requires non-optional options to be passed', () => {
      expect(
        () =>
          // @ts-ignore Allow missing prop
          new RequiredProps(),
      ).toThrowErrorMatchingSnapshot();
    });
  });

  describe('configure()', () => {
    it('can set an option', () => {
      const opts = new OptionalProps();

      expect(opts.options.foo).toBe('default');

      opts.configure({
        foo: 'new',
      });

      expect(opts.options.foo).toBe('new');
    });

    it('persists other options not being configured', () => {
      const opts = new OptionalProps({
        foo: 'abc',
        bar: 123,
      });

      expect(opts.options).toEqual({
        foo: 'abc',
        bar: 123,
      });

      opts.configure({
        bar: 456,
      });

      expect(opts.options).toEqual({
        foo: 'abc',
        bar: 456,
      });
    });
  });
});
