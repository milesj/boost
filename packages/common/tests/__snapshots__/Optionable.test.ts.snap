// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Optionable constructor() errors for invalid option type passed 1`] = `"Invalid OptionalProps field \\"foo\\". Must be a string."`;

exports[`Optionable constructor() errors for unknown option 1`] = `"Unknown fields: unknown."`;

exports[`Optionable constructor() requires non-optional options to be passed 1`] = `"Invalid RequiredProps field \\"bar\\". Field is required and must be defined."`;
